### DISCUSSION TITLE
How to convert directly  collected packets into flows instead of first saving them as a pcap file

### DISCUSSION THREAD

**Discussion by DarmJarlo** [nfstream/nfstream#167] [Q&A]
Hey I am trying to collect some packets through ryu controller and then convert them into flows with NFstream.The whole process is like first I collect the packets, save them in a list and write them in a pcap file, and then use NFstream to transform this pcap file. My idea is to convert this list of packets directly into flows with NFstream without saving these pcap. Because writing in this pcap file on the ryu controller take too much time. Could someone help？Thanks in advance

By the way, I am reading the source code and have come up with a primary idea that I can modify your lib_engine.c to save the list of packets in pcap file as you do for the interface mode. But I cannot find the code about the function pcap_create in line 1449. Can you help to tell me where is its definition? Thanks!

**Comment by drnpkr** [comment]
Hi @DarmJarlo,

You can deploy custom applications on top of the controller using the northbound interface. This allows you to use NFStream in live mode where it can convert all the packets into flows and store them in a CSV or forward them somewhere else (messaging broker) for analysis. This certainly requires some implementation on your side. 

However, I do not understand your strategy to collect packets through the controller. Packets are being sent there only if there is no flow table entry (unless you configure the data plane to forward all the packets to the controller). Once there is a flow table entry, upcoming packets of a flow won't be forwarded to the controller. This way you miss packets for certain flows. Am I missing something reagarding your approach?

**Reply by DarmJarlo** [reply]
Hey, thanks for your quick reply!
Actually I have deployed this application on top of the ryu controller. But the problem is NFstream only allows us to input pcap file or network interface as the source. As mentioned, pcap file is not appropriate here. And because I want to select some packets before inputting to Nfstream, to directly monitor an interface and get all the packets together is not what I need as well. The approach I have so far is to store the selected packets in a list ( implemented as : 

self.address_pair_packetlist[addr_pair].append(ev.msg.data) #ev.msg.data is the packet

and then input this list to NFstream. So the main challenge here is how to let NFstream handle these packets in a list as it handles the packets in a pcap file.

Thanks in advance!

**Comment by drnpkr** [comment]
@DarmJarlo, 

Can you plase help me better understand your goal? I think, your approach is flawed.

An SDN controller is not suitable for flow measurement. Assuming you are doing this for research, if I were a reviewer, I would question such a proposal because of the placement of the measurement. Now your are sending all your traffic to the cloud just because you need to pull out some statistics. Packet capture and flow measurement should take place in the data plane.

Is there a specific reason for such a placement?
